[View in Colaboratory](https://colab.research.google.com/github/Bhatnagar-Ashish/test/blob/master/Assignment2B.ipynb)

## Step 1: Read input and output



In [36]:
import numpy as np

def sigmoid(x):   # Define sigmoid function.
   return 1 / (1 + np.exp(-x))
  
def slope(x):      # Define Slope function.
    return x*(1-x)
  
x=np.array([[1,0,0,0],[0,0,0,1],[0,1,1,0]], dtype=int)
print("Input: ")
print(x)

y=np.array([[1],[0],[0]])
print("Y: ")
print(y)

Input: 
[[1 0 0 0]
 [0 0 0 1]
 [0 1 1 0]]
Y: 
[[1]
 [0]
 [0]]


### Input 

$$\mathbf{x} = \left[\begin{array}
{rrr}
1  & 0 & 0 & 0 \\
0 & 0 & 0 & 1 \\
0 & 1 & 1 & 0
\end{array}\right]
$$


### Output

$$\mathbf{y} = \left[\begin{array}
{rrr}
1 \\
0 \\
0
\end{array}\right]
$$

## Step 2: Generate random weights and biases


In [22]:


weight = np.around(np.random.random((4,3)),decimals=2)  # Create an 4X3 array filled with random values
print("Weight: ")
print(weight) 

bias=np.around(np.random.random((1,3)),decimals=2)  # Create an 1X3 array filled with random values
print("Bias: ")
print(bias) 

Weight: 
[[0.17 0.61 0.61]
 [0.28 0.18 0.26]
 [0.23 0.82 0.32]
 [0.83 0.58 0.77]]
Bias: 
[[0.68 0.64 0.73]]


### Weight 

$$\mathbf{weight} = \left[\begin{array}
{rrr}
0.17  & 0.61 & 0.61 \\
0.28 & 0.18 & 0.26 \\
0.23 & 0.82 & 0.32 \\
0.83 & 0.58 & 0.77
\end{array}\right]
$$


### Bias

$$\mathbf{bias} = \left[\begin{array}
{rrr}
0.68 & 0.64 & 0.73
\end{array}\right]
$$

In [25]:
import numpy as np

w_out = np.around(np.random.random((3,1)),decimals=2)  # Create an 3X1 array filled with random values
print("Weight Out: ")
print(w_out) 

b_out=np.around(np.random.random(),decimals=2)  # Create variable with random values
print("Bias Out: ")
print(b_out) 

Weight Out: 
[[0.22]
 [0.12]
 [0.85]]
Bias Out: 
0.46


### Weight Out 

$$\mathbf{w\_out} = \left[\begin{array}
{rrr}
0.22  \\
0.12 \\
0.85 
\end{array}\right]
$$


### Bias Out

$$\mathbf{b\_out} = \left[\begin{array}
{rrr}
0.46 
\end{array}\right]
$$

## Step 2: Calculate hidden layer input

>hidden_layer_input = matrix_dot_product(X,wh) + bh



In [31]:


from numpy.core.multiarray import ndarray

hidden_layer_input: ndarray = x.dot(weight) + bias
print(hidden_layer_input)



[[0.85 1.25 1.34]
 [1.51 1.22 1.5 ]
 [1.19 1.64 1.31]]


### Hidden layer input
$$\mathbf{hidden\_layer\_input} = \left[\begin{array}
{rrr}
0.85 & 1.25 & 1.34  \\
1.51 & 1.22 & 1.5 \\
1.19 & 1.64 & 1.31
\end{array}\right]
$$


## Step 3: Perform non-linear transformation on hidden linear input
>hidden_layer_activations = sigmoid(hidden_layer_input)




In [33]:
hidden_layer_activations = sigmoid(hidden_layer_input)
print(hidden_layer_activations)

[[0.7  0.78 0.79]
 [0.82 0.77 0.82]
 [0.77 0.84 0.79]]


### Hidden layer activations
$$\mathbf{hidden\_layer\_activations} = \left[\begin{array}
{rrr}
0.7 & 0.78& 0.79  \\
0.82 & 0.77 & 0.82 \\
0.77 & 0.84 & 0.79
\end{array}\right]
$$

## Step 4: Perform linear and non-linear transformation of hidden layer activation at output layer
>output_layer_input = matrix_dot_product (hidden_layer_activations * wout ) + bout 

>output = sigmoid(output_layer_input)




In [34]:
output_layer_input = hidden_layer_activations.dot(w_out) + b_out
output=sigmoid(output_layer_input)
print("Output: ")
print(output)

Output: 
[[0.8 ]
 [0.81]
 [0.8 ]]


### Output
$$\mathbf{output} = \left[\begin{array}
{rrr}
0.8  \\
0.81 \\
0.8
\end{array}\right]
$$

## Step 5: Calculate gradient of Error(E) at output layer
>E = y-output




In [35]:
E= y - output
print(E)



[[ 0.2 ]
 [-0.81]
 [-0.8 ]]


### Error
$$\mathbf{E} = \left[\begin{array}
{rrr}
0.2  \\
-0.81 \\
-0.8
\end{array}\right]
$$

## Step 6: Compute slope at output and hidden layer

> Slope_output_layer= derivatives_sigmoid(output)

> Slope_hidden_layer = derivatives_sigmoid(hiddenlayer_activations)`


In [41]:
slope_output=slope(output)
print("Slope Output")
print(slope_output)

slope_hidden_layer=slope(hidden_layer_activations)
print("Slope Hidden Layer")
print(slope_hidden_layer)

Slope Output
[[0.16]
 [0.16]
 [0.16]]
Slope Hidden Layer
[[0.21 0.17 0.16]
 [0.15 0.18 0.15]
 [0.18 0.14 0.17]]


### Slope at output

$$\mathbf{slope\_output} = \left[\begin{array}
{rrr}
0.16 \\
0.16 \\
0.16
\end{array}\right]
$$


### Slope at hidden layer

$$\mathbf{slope\_hidden\_layer} = \left[\begin{array}
{rrr}
0.21  & 0.17 & 0.16 \\
0.15 & 0.18 & 0.15 \\
0.18 & 0.14 & 0.17 
\end{array}\right]
$$

## Step 7: Compute delta at output layer


> d_output = E \* slope_output_layer\*lr


In [42]:
lr = 0.1

d_output = E * Slope_output * lr
print(d_output)



[[ 0.  ]
 [-0.01]
 [-0.01]]


### Delta at output

$$\mathbf{d\_output} = \left[\begin{array}
{rrr}
0 \\
-0.01 \\
-0.01
\end{array}\right]
$$

## Step 8: Calculate Error at hidden layer


> Error_at_hidden_layer = matrix_dot_product(d_output, wout.Transpose)



In [43]:
error_at_hidden_layer=d_output * w_out.transpose()
print(error_at_hidden_layer)



[[ 0.    0.    0.  ]
 [-0.   -0.   -0.01]
 [-0.   -0.   -0.01]]


### Error at hidden layer

$$\mathbf{error\_at\_hidden\_layer} = \left[\begin{array}
{rrr}
0  & 0 & 0 \\
0 & 0 & -0.1 \\
0 & 0 & -0.1 
\end{array}\right]
$$

## Step 9: Compute delta at hidden layer


> d_hiddenlayer = Error_at_hidden_layer * slope_hidden_layer





In [44]:
d_hidden_layer=error_at_hidden_layer*slope_hidden_layer
print(d_hidden_layer)

[[ 0.  0.  0.]
 [-0. -0. -0.]
 [-0. -0. -0.]]


### Delta at hidden layer

$$\mathbf{d\_hidden\_layer} = \left[\begin{array}
{rrr}
0  & 0 & 0 \\
0 & 0 & 0 \\
0 & 0 & 0 
\end{array}\right]
$$

## Step 10: Update weight at both output and hidden layer


> wout = wout + matrix_dot_product (hiddenlayer_activations.Transpose, d_output) * learning_rate

> wh = wh+ matrix_dot_product (X.Transpose,d_hiddenlayer) * learning_rate

In [48]:
w_out=w_out+(hidden_layer_activations.transpose().dot(d_output)) * lr
print("w_out:")
print(w_out)

weight=weight+x.transpose().dot(d_hidden_layer) * lr
print("Weight:")
print(weight)



w_out:
[[0.22]
 [0.12]
 [0.85]]
Weight:
[[0.17 0.61 0.61]
 [0.28 0.18 0.26]
 [0.23 0.82 0.32]
 [0.83 0.58 0.77]]


### Updated weight and w_out

$$\mathbf{w\_out} = \left[\begin{array}
{rrr}
0.22\\
0.12 \\
0.85 
\end{array}\right]
$$

$$\mathbf{weight} = \left[\begin{array}
{rrr}
0.17  & 0.61 & 0.61 \\
0.28 & 0.18 & 0.26 \\
0.23 & 0.82 & 0.32 \\
0.83 & 0.58 & 0.77
\end{array}\right]
$$


## Step 11: Update biases at both output and hidden layer


> bh = bh + sum(d_hiddenlayer, axis=0) * learning_rate

> bout = bout + sum(d_output, axis=0)*learning_rate

In [49]:
bias = bias + np.sum(d_hidden_layer, axis=0) * lr
print("Bias:")
print(bias)
print("Bias Out:")
b_out = b_out + np.sum(d_output, axis=0) * lr
print(b_out)

Bias:
[[0.68 0.64 0.73]]
Bias Out:
[0.46]



### Bias

$$\mathbf{bias} = \left[\begin{array}
{rrr}
0.68 & 0.64 & 0.73
\end{array}\right]
$$

### Bias Out

$$\mathbf{b\_out} = \left[\begin{array}
{rrr}
0.46 
\end{array}\right]
$$